pyleros/fedec.py

```diff
@@ -0,0 +1,198 @@
+            an direct add or and indirect load/ store(which follows)
+        pipe_dec: OUT List of the decode signals, pass on to the execute stage
+        pipe_imme: OUT Immediate value, as taken from the lower bits 
+                of the instruction, pass on to execute stage
+        pipe_dm_addr: OUT DM read addr, pipeline register
+        pipe_pc: OUT the value of PC, pipeline register
+
+    Parameters:
+
+        filename: Name of the file or a list containing the instructions
+
+    """
+
+    im_addr = Signal(intbv(0)[IM_BITS:])
+
+    instr = Signal(intbv(0)[16:])
+    instr_hi = Signal(intbv(0)[8:])
+
+    branch_en, acc_z = False, True
+
+    pc, pc_next = [Signal(intbv(0)[IM_BITS:])] * 2
+    pc_add = intbv(0)[IM_BITS:]
+
+    decode = [Signal(bool(0)) for i in dlist]
+    decode[int(t_decSignal.op)] = Signal(alu_op_type.LD)
+
+    # Instantiate the instruction memory
+    im_inst = rom.pyleros_im(clk, reset, im_addr, instr, filename=filename)
+
+    # Instantiate the decoder
+    dec_inst = decoder.pyleros_decoder(instr_hi, decode)
+
+    @always_comb
+    def sync_sig():
+
+        # if not reset == reset.active:
+        instr_hi.next = instr[16:8]
+        im_addr.next = pc_next
+        pipe_pc.next = pc_add
+
+
+    @always_comb
+    def dm_addr_sel():
+
+        # if not reset == reset.active:
+        # Note ind ls adds the 8 bit immediate offset
+        # to the previous value of the dm_
+        offset_addr = intbv(dm_data + instr[8:0])[16:]
+        
+        # Indirect Addressing(with offset) 
+        # for indirect load/store
+        if decode[int(t_decSignal.indls)]:
+            pipe_dm_addr.next = offset_addr[DM_BITS:] 
+
+        # Direct Addressing
+        else:
+            pipe_dm_addr.next = instr[DM_BITS:]
+
+    @always_comb
+    def branch_sel():
+
+        # if not reset == reset.active:
+        if acc == 0:
+            acc_z = True
+
+        else:
+            acc_z = False
+
+        branch_en = 0
+
+        if decode[int(t_decSignal.br_op)]:
+            br_type = instr[11:8]
+
+            if br_type == 0b000:
+                # BRANCH
+                branch_en == 1
+
+            elif br_type == 0b001:
+                # BRZ
+                branch_en = True if acc_z else False
+
+            elif br_type == 0b010:
+                # BRNZ
+                branch_en = True if not acc_z else False
+
+            elif br_type == 0b011:
+                # BRP
+                branch_en = True if not acc[15] else False
+
+            elif br_type == 0b100:
+                # BRN
+                branch_en = True if acc[15] else False
+
+    # For selection of next PC address
+    @always_comb
+    def pc_mux():
+
+        pc_add_tmp = intbv(0)[IM_BITS:]
+        pc_op = intbv(0)[IM_BITS:]
+        # if not reset == reset.active:
+        if branch_en:
+            # Sign extend the low 8 bits
+            # of instruction
+            pc_op[:] = sign_extend(instr, IM_BITS)
```

Comment:
The `sign_extend` function is certainly not convertible, so by extension neither is this.

I suggest you do this using an intermediate block.